### PR TITLE
fix Libc.rand and seed problems

### LINKED ISSUE #43704: new randomdevice code needs unittests
Can you add a unittest for this (just literally call `@test_warn "ad-hoc" Base._ad_hoc_entropy_source()`, and make sure it returns a reasonable value? and related PR comments.

_Originally posted by @vtjnash in https://github.com/JuliaLang/julia/pull/43606#discussion_r780451488_

### PR DESCRIPTION
**Pull request by vtjnash**
Continuation from #43606
Close #43704

 - Replaces thread-unsafe function `rand` with `jl_rand`.
 - Fixes `_ad_hoc_entropy_source` fallback in Random.
 - Uses uv_random for more direct access to quality-randomness (usually
   a syscall rather than a file.)
 - Ensures Array{Bool} are valid when created from RandomDevice.

### CODE REVIEW

**Review by aviatesk — COMMENTED** [review]
I read through the changes and couldn't find any obvious issues -- still probably I am not familiar enough with these code bases to give an approval.

**Review by staticfloat — APPROVED** [review]
Seems reasonable to me, just a few questions around external visibility of the removed symbol

File: src/jl_exported_funcs.inc [hunk 1]
```diff
@@ -201,7 +201,6 @@
     XX(jl_getallocationgranularity) \
     XX(jl_getnameinfo) \
     XX(jl_getpagesize) \
-    XX(jl_getpid) \
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply src/jl_exported_funcs.inc:204]: Our C API is not declared stable

File: src/julia_internal.h [hunk 2]
```diff
@@ -1159,7 +1157,6 @@ JL_DLLEXPORT const char *jl_dlfind_win32(const char *name);
 
 // libuv wrappers:
 JL_DLLEXPORT int jl_fs_rename(const char *src_path, const char *dst_path);
-int jl_getpid(void) JL_NOTSAFEPOINT;
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply src/julia_internal.h:1162]: also true that it wasn't accessible

File: src/init.c [hunk 3]
```diff
@@ -685,11 +685,12 @@ JL_DLLEXPORT void julia_init(JL_IMAGE_SEARCH rel)
         jl_error("cannot generate code-coverage or track allocation information while generating a .o, .bc, or .s output file");
     }
 
+    jl_init_rand();
     jl_init_runtime_ccall();
-    jl_gc_init();
     jl_init_tasks();
     jl_init_threading();
 
+    jl_gc_init();
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply src/init.c:693]: seemed slightly more consistent, since eventually threads will start up later, and the GC needs to be configured earlier

**Pull request merged**